pkg/containerd/events.go

```diff
@@ -253,6 +267,98 @@ func (e *EventMonitor) updateListener(
 	}
 }
 
+// execIptablesRules creates an additional DNAT rule if port binding is bound to 127.0.0.1.
+func execIptablesRules(portMappings nat.PortMap, containerID, namespace, pid string) error {
+	var errs []error
+
+	for portProto, portBindings := range portMappings {
+		for _, portBinding := range portBindings {
+			if portBinding.HostIP == "127.0.0.1" {
+				err := createLoopbackIPtablesRules(containerID, namespace, pid, portProto.Port(), portBinding.HostPort)
+				if err != nil {
+					errs = append(errs, err)
+				}
+			}
+		}
+	}
+
+	if len(errs) != 0 {
+		return fmt.Errorf("%w: %+v", ErrExecIptablesRule, errs)
+	}
+
+	return nil
+}
+
+// When namespaced network is enabled, if the port binding is bound to 127.0.0.1, we add an additional
+// DNAT rule in the main CNI DNAT chain (CNI-HOSTPORT-DNAT) after the existing rule (using --append).
+// This is necessary because the initial CNI rule created by containerd only allows the traffic
+// to be routed to localhost. Therefore, we add an additional rule to allow traffic to any
+// destination IP address which allows the service to be discoverable through namespaced network's
+// subnet, which essentially causes the service to listen on eth0 instead; this is required as the
+// traffic is routed via vm-switch over the tap network.
+// The existing DNAT rule are as follows:
+// DNAT       tcp  --  anywhere             localhost            tcp dpt:9119 to:10.4.0.22:80.
+// We enter the following rule after the existing rule:
+// DNAT       tcp  --  anywhere             anywhere             tcp dpt:9119 to:10.4.0.22:80.
+func createLoopbackIPtablesRules(containerID, namespace, pid, port, destinationPort string) error {
+	// read the container's cni config to extract the network name
+	nsenterNetworkConfCmd := exec.Command("nsenter", "-t", pid, "-n", "cat", "/etc/cni/net.d/nerdctl-bridge.conflist")
+	output, err := nsenterNetworkConfCmd.CombinedOutput()
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("read the following network conflist for [containerID: %s] config: %s", containerID, string(output))
+
+	var networkConfig cniNetworkConfig
+	err = json.Unmarshal(output, &networkConfig)
+	if err != nil {
+		return err
+	}
+
+	eth0IP, err := extractIPAddress(pid)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("found the ip address: %s for containerID: %s", eth0IP, containerID)
+
+	// create the corresponding chanin name, e.g CNI-DN-xxxxxx
+	// (where xxxxxx is a function of the ContainerID and network name)
+	// https://www.cni.dev/plugins/current/meta/portmap/#dnat
+	// https://github.com/containernetworking/plugins/blob/2b097c5a62dacfd2f9ea4268faa4fc04bd5343f5/pkg/utils/utils.go#L39
+	cID := fmt.Sprintf("%s-%s", namespace, containerID)
+	chainName := mustFormatHashWithPrefix(maxChainLength, chainPrefix+"DN-", networkConfig.Name+cID)
```

Comment:
It's odd that `chainPrefix` is a constant that's only used here, concatenated with a second string. But we can fix that later.